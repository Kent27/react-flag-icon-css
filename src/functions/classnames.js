// @flow
import classnames from 'classnames'
import type classes from 'classnames'
import type { FlagIconPropsType, FlagIconOptionsType } from '../types/flow'
import { baseThemeStyleName, flagIconClassesPrefix, flagIconClassesPrefixName } from './constants'


export const makeClassesObject = ({ ...props,
                                       code,
                                       flip = '',
                                       size = '',
                                       squared = false,
                                       rotate }: FlagIconPropsType,
                                       options: FlagIconOptionsType) : classes => {
  let obj = {
    [flagIconClassesPrefixName]: true,
    [`${flagIconClassesPrefix}squared`]: squared,
    [`${flagIconClassesPrefix}${size}`]: size,
    [`${flagIconClassesPrefix}${code}`]: code,
    [`${flagIconClassesPrefix}flip-${flip}`]: flip,
    [`${flagIconClassesPrefix}rotate-${rotate ? rotate : ''}`]: rotate, // eslint-disable-line no-unneeded-ternary
    [baseThemeStyleName]: options.useCssModules && options.themeStyles
  }

  if (props.className && options.useCssModules) {
    obj = { ...obj, [props.className]: true }
  }
  return obj
}

export default (props: FlagIconPropsType, options: FlagIconOptionsType) : string =>
  classnames(makeClassesObject(props, options))
