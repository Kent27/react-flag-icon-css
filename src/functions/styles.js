// @flow
import type { FlagIconOptionsType, ObjStringKeyValuesType } from '../types/flow'


export default (
  styles: ObjStringKeyValuesType,
  options?: FlagIconOptionsType
): ObjStringKeyValuesType => { // eslint-disable-line arrow-body-style
  return (options && options.themeStyles) ? { ...styles, ...options.themeStyles } : styles
}
