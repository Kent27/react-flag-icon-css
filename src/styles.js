import stylesFlagIcon from 'flag-icon-css/css/flag-icon.css'
import stylesMain from './styles/main.css'

const finalStyles = {
  ...stylesFlagIcon,
  ...stylesMain,
}

export default finalStyles
