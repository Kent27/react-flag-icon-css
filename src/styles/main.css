:root {
  --font-sizes: lg, 2x, 3x, 4x, 5x; /* stylelint-disable-line unit-no-unknown */
  --font-sizes-values: 1.33333333em, 2em, 3em, 4em, 5em;
  --rotates: 30, 60, 90, 180, 270;
}

.flag-icon-flip-horizontal {
  transform: scale(-1, 1);
}

.flag-icon-flip-vertical {
  transform: scale(1, -1);
}

@each $key, $val in (var(--font-sizes)), (var(--font-sizes-values)) {
  .flag-icon-$key {
    font-size: $val;
  }
}

@each $key in (var(--rotates)) {
  .flag-icon-rotate-$key {
    transform: rotate($(key)deg); /* stylelint-disable-line function-whitespace-after */
  }
}
