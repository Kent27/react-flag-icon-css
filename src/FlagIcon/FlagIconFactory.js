// @flow
import CssModulesTransform from 'react-css-modules'

import typeof ReactModule from 'react'
import FlagIcon from './FlagIcon'
import styles from '../styles'
import { makeStyles, makeFlagIconOptions } from '../functions'
import { FlagIconPropsType as FlagIconPropsTypeReact } from '../types/propTypes'
import type {
  FlagIconOptionsType,
  FlagIconFactoryReturnType,
} from '../types/flow'

const FlagIconFactory = (
  React: ReactModule,
  options?: FlagIconOptionsType,
): FlagIconFactoryReturnType => {
  // We 'makeFlagIconOptions' by merging the default options with the (optional)
  // user-supplied options.
  const computedOptions = makeFlagIconOptions(options)
  const { useCssModules } = computedOptions

  // We assign the React Component to 'FlagIconComponent'. SEE: './FlagIcon.js'
  const FlagIconComponent = FlagIcon(React, computedOptions)
  FlagIconComponent.displayName = 'FlagIcon' // Otherwise 'WrappedComponent' when testing.

  if (process.env.NODE_ENV !== 'production') {
    // UglifyJS strips this block out in production.
    // We assign react propTypes (dynamic type checking) to the React component.
    // SEE: https://github.com/reactjs/prop-types
    FlagIconComponent.propTypes = FlagIconPropsTypeReact
  } // In production, FlagIconComponent.propTypes will be undefined.

  if (useCssModules) {
    const computedStyles = makeStyles(styles, options)

    if (process.env.NODE_ENV !== 'production') {
      // UglifyJS strips this block out in production.
      if (useCssModules) {
        // Has the user set 'useCssModules' to true or is it just the default value?
        const userSetUseCssModules = options ? options.useCssModules : false

        if (
          !Object.prototype.hasOwnProperty.call(computedStyles, 'flag-icon')
        ) {
          const readme =
            'https://github.com/matteocng/react-flag-icon-css/blob/master/README.md'
          const readmeModules =
            'https://github.com/gajus/react-css-modules#usage'

          // eslint-disable-next-line no-console
          console.error(`Warning: Detected configuration mismatch/error: [react-flag-icon-css] \
FlagIconFactory tried to build FlagIcon with the option 'useCssModules' set to true \
(${userSetUseCssModules ? 'by you' : 'default'}), but it seems that your environment \
is not loading the stylesheets in the appropriate way. Two possibilities: \
\n\n 1. Your project is not using CSS Modules / you don't want to use them. \
Try building FlagIcon like this: \n\n\tFlagIconFactory(React, { useCssModules: false }) \
\n\n\tSee the README for more info: ${readme} \
\n\n 2. You are using CSS Modules but your module bundler e.g Webpack \
is not correctly set up for CSS modules. If you are using Webpack, the key \
modules to set up are 'style-loader', 'css-loader' and 'extract-text-webpack-plugin' \
('sass-loader' also comes into play if you are using SASS). \
The link below may help you:\n\n\t${readmeModules}`)
        }
      }
    }

    /*
     *  We decorate the 'FlagIconComponent' using React Css Modules' decorator function.
     *
     *  SEE:  https://github.com/gajus/react-css-modules#decorator
     *        https://github.com/gajus/react-css-modules#options
     *
     *  When the 'FlagIconComponent' built by this factory is instantiated with some
     *  props, CssModulesTransform may throw a '... CSS module is undefined' Error.
     *
     *  It can be caused by:
     *
     *  -  an incorrect configuration of the user's module bundler (Webpack
     *     loaders...), meaning that the styles have not been loaded in the format
     *     that 'react-css-modules' needs and expects.
     *
     *     SEE: https://github.com/gajus/react-css-modules#usage
     *
     *  -  options.useCssModules passed as 'true' to this factory (or 'true' as
     *     default), but the user's module bundler is configured for standard CSS
     *     classes and not CSS modules.
     *
     *  -  invalid props supplied to 'FlagIconComponent'; a side effect is that the
     *     'styleName' property of 'FlagIconComponent', which is a function of its
     *     props and the options it was built with, contains CSS modules not part of
     *     'computedStyles'. If NODE_ENV !== 'production', 'prop-types' prints a debugging
     *     message to the browser console (before 'CssModulesTransform' is executed).
     *
     *  -  'options.themeStyles' was passed and 'options.useCssModules' is
     *     'true'. The user then passed a custom code e.g 'z3' as a prop to
     *     'FlagIconComponent' but he/she forgot to include .flag-icon-z3 { ... }
     *     in the stylesheet/s loaded-mapped to 'themeStyles'.
     *
    */
    return CssModulesTransform(FlagIconComponent, computedStyles, {
      allowMultiple: true,
    })
  } // /if useCssModules

  return FlagIconComponent
}

export default FlagIconFactory
