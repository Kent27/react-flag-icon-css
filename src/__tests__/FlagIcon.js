// @flow
import React from 'react'
import test from 'tape' // eslint-disable-line import/no-extraneous-dependencies
import { shallow, mount } from 'enzyme' // eslint-disable-line import/no-extraneous-dependencies
import classnames from 'classnames'

import type { ReactWrapper, ShallowWrapper } from 'enzyme' // eslint-disable-line import/no-extraneous-dependencies
import type classes from 'classnames'
import type { FlagIconPropsType, FlagIconOptionsType } from '../types/flow'

import FlagIconFactory from '../'
import styles from '../styles'
import DummyComponentFactory from './DummyComponent'
import { countries, constants, makeStyles, makeClassesObject, objectKeysApplyFn,
         makeFlagIconOptions } from '../functions'
import testThemeStyles from './testThemeStyles.scss'

const { getCountryCodes } = countries
const { flagIconClassesPrefixName } = constants

// Helpers and definitions

const getExpectedClassName = (props: FlagIconPropsType, options: FlagIconOptionsType) : string => {
  const computedOptions = makeFlagIconOptions(options)
  let oParams = makeClassNameObject(props, computedOptions)

  if (computedOptions.useCssModules) {
    const computedStyles = makeCssModulesObject(styles, computedOptions)

    oParams = objectKeysApplyFn(oParams, (key: string) : string => {
      const skipValue = !oParams[key] // keys with 'false' values can be ignored
      const newKeyName = computedStyles[key]
      return skipValue ? key : newKeyName
    })
  }
  return classnames(oParams)
}

const filterClassObjectKey = (obj: classes, objKey: string) : boolean =>
  objKey !== flagIconClassesPrefixName && objKey.startsWith(flagIconClassesPrefixName)

// FlagIcon Props

const requiredProps = {
  code: 'it'
}

const optionalProps = {
  size: '3x',
  squared: false,
  rotate: 90,
  flip: 'horizontal'
}

const allProps = {
  ...requiredProps,
  ...optionalProps
}

// Tests

test('functions > makeClassesObject', (t: tape$Context) => {
  const computedOptions = makeFlagIconOptions()
  const oParams = makeClassesObject(allProps, computedOptions)

  // First remove ${flagIconClassesPrefixName}, any key not starting with
  // ${flagIconClassesPrefixName} (e.g theme key)
  const oFilteredParams = Object.keys(oParams).filter(filterClassObjectKey.bind(this, oParams))
  const expectedLength = Object.keys(optionalProps).length
  // We substract the length of the required properties and then expect a key in the
  // object created by makeClassesObject for each of the optional properties
  const computedLength = oFilteredParams.length - Object.keys(requiredProps).length

  t.equal(expectedLength, computedLength)
  t.end()
})

test('FlagIconFactory > useCssModules: false and props: className', (t: tape$Context) => {
  const options = { useCssModules: false }
  const FlagIcon = FlagIconFactory(React, options)
  const props = { ...baseProps, className: 'some-css-rule' }

  // Test that className works when useCssModules: false
  const ReactFlagIcon : React$Element<*> = FlagIcon(props)
  const wrapper: ShallowWrapper<*> = shallow(ReactFlagIcon)
  const expectedClassName = getExpectedClassName(props, options)

  t.equal(wrapper.contains(<span className={expectedClassName} />), true)
  t.end()
})

test('FlagIconFactory', (t: tape$Context) => {
  FlagIconFactory(React)
  t.end()
})

test('FlagIcon themeStyles', (t: tape$Context) => {
  const options = { themeStyles: testThemeStyles }
  const FlagIconCssModules = FlagIconFactory(React, options)
  const ReactFlagIcon = FlagIconCssModules({ ...baseProps })

  const wrapper: ShallowWrapper<*> = shallow(ReactFlagIcon)
  const expectedClassName = getExpectedClassName(baseProps, options)

  t.equal(wrapper.contains(<span className={expectedClassName} />), true)
  t.end()
})

test('FlagIcon props', (t: tape$Context) => {
  const options = { themeStyles: testThemeStyles }
  const FlagIconCssModules = FlagIconFactory(React, options)
  const FlagIcon = FlagIconFactory(React, makeFlagIconOptions({ useCssModules: false }))
  const props = {
    size: '3x',
    squared: '',
    rotate: 180
  }

  let currentProps = { ...baseProps }
  let currentExpectedClassName = ''

  Object.keys(props).forEach((prop: string) => {
    const ReactFlagIconCssModules : React$Element<*> = FlagIconCssModules({ ...currentProps })
    const ReactFlagIcon : React$Element<*> = FlagIcon({ ...currentProps })
    const FlagIcons = [ReactFlagIconCssModules, ReactFlagIcon]
    FlagIcons.forEach((flagIcon: React$Element<*>, i: number) => {
      const wrapper: ShallowWrapper<*> = shallow(flagIcon)
      const Component = currentProps.Component ? currentProps.Component : 'span'
      const currentOptions = { ...options, useCssModules: (i === 0) }
      currentExpectedClassName = getExpectedClassName(currentProps, currentOptions)

      t.equal(wrapper.contains(<Component className={currentExpectedClassName} />), true)
    })
    currentProps = { ...currentProps, [prop]: props[prop] }
  })
  t.end()
})

test('FlagIcon props:Component', (t: tape$Context) => {
  const options = { themeStyles: testThemeStyles }
  const FlagIconCssModules = FlagIconFactory(React, options)
  const FlagIcon = FlagIconFactory(React, { useCssModules: false })
  const ReactFlagIconCssModules = FlagIconCssModules({ ...baseProps, Component: 'div' })
  const ReactFlagIcon = FlagIcon({ ...baseProps, Component: 'div' })
  const flagIcons = [ReactFlagIconCssModules, ReactFlagIcon]
  let currentExpectedClassName = ''

  flagIcons.forEach((flagIcon: React$Element<*>, i: number) => {
    const wrapper: ShallowWrapper<*> = shallow(flagIcon)
    const currentOptions = { ...options, useCssModules: (i === 0) }
    currentExpectedClassName = getExpectedClassName(baseProps, currentOptions)

    t.equal(wrapper.contains(<div className={currentExpectedClassName} />), true)
  })
  t.end()
})

test('FlagIcon mount > props:children', (t: tape$Context) => {
  const FlagIconCssModules = FlagIconFactory(React, { themeStyles: testThemeStyles })
  const FlagIcon = FlagIconFactory(React, { useCssModules: false })
  const childrenClassName = 'test'
  const childrenText = 'test'
  const children = DummyComponentFactory(React)({ text: childrenText,
                                                  className: childrenClassName })
  const ReactFlagIconCssModules = FlagIconCssModules({ ...baseProps, children })
  const ReactFlagIcon = FlagIcon({ ...baseProps, children })
  const flagIcons = [ReactFlagIconCssModules, ReactFlagIcon]

  flagIcons.forEach((flagIcon: React$Element<*>) => {
    const wrapper: ReactWrapper<*> = mount(flagIcon)

    t.equal(wrapper.contains(children), true)
    t.equal(wrapper.find('div').is(`div.${childrenClassName}`), true)
    t.equal(wrapper.find('div').text(), childrenText)
  })
  t.end()
})
