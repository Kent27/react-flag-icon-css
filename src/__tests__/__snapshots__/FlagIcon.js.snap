// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optional props 1`] = `
<span
  className="flag-icon flag-icon-it"
/>
`;

exports[`optional props 2`] = `
<span
  className="flag-icon flag-icon-3x flag-icon-it"
/>
`;

exports[`optional props 3`] = `
<span
  className="flag-icon flag-icon-3x flag-icon-it"
/>
`;

exports[`optional props 4`] = `
<span
  className="flag-icon flag-icon-3x flag-icon-it flag-icon-rotate-90"
/>
`;

exports[`options:themeStyles 1`] = `
<span
  className="__MODULE__flag-icon __MODULE__flag-icon-it __MODULE__theme-base"
/>
`;

exports[`props: className 1`] = `
<span
  className="flag-icon flag-icon-it some-css-rule"
/>
`;

exports[`props: className, props: styleName 1`] = `
<span
  className="flag-icon flag-icon-it some-css-rule"
/>
`;

exports[`props: className, props: styleName 2`] = `
<span
  className="__MODULE__flag-icon __MODULE__flag-icon-it __MODULE__theme-base __MODULE__some-other-css-rule __MODULE__some-css-rule"
/>
`;

exports[`props: styleName 1`] = `
<span
  className="__MODULE__flag-icon __MODULE__flag-icon-it __MODULE__theme-base __MODULE__some-other-css-rule __MODULE__some-css-rule"
/>
`;

exports[`props:Component 1`] = `
<div
  className="flag-icon flag-icon-it"
/>
`;

exports[`props:Component 2`] = `
<div
  className="__MODULE__flag-icon __MODULE__flag-icon-it"
/>
`;

exports[`props:children 1`] = `
<span
  className="flag-icon flag-icon-it"
>
  <div
    className="test"
  >
    test
  </div>
</span>
`;

exports[`props:children 2`] = `
<span
  className="__MODULE__flag-icon __MODULE__flag-icon-it"
>
  <div
    className="test"
  >
    test
  </div>
</span>
`;
